File: 프로그래머스/lv2/17680. ［1차］ 캐시/［1차］ 캐시.cpp
#include <string>
#include <vector>
#include <list>
#include <algorithm>

using namespace std;

int solution(int cache_size, vector<string> cities)
{
	list<string> cache;
	list<string>::iterator it;
	int times = 0;

	for (string& city : cities)
	{
		for_each(city.begin(), city.end(), [](char& c) {c = tolower(c); });
        
        it = find(cache.begin(), cache.end(), city);
        if(it != cache.end()) // cache hit
        {
            times += 1;
            it = cache.erase(it);
            cache.emplace_front(city);
        }
        else // cache miss
        {
            times += 5;
            cache.emplace_front(city);
		    if (cache.size() > cache_size) 
                cache.pop_back();
        }
	}

	return times;
}

File: 프로그래머스/2/17680. ［1차］ 캐시/［1차］ 캐시.py
from collections import deque

HIT = 1
MISS = 5

def solution(cache_size, cities):
    cache = deque()
    time = 0
    
    for city in cities:
        city = city.lower()
        
        if city in cache:
            time += HIT
            cache.remove(city)
            cache.append(city)
        else:
            time += MISS
            cache.append(city)
            if len(cache) > cache_size:
                cache.popleft()
        
    return time